### PR TITLE
Implement -assert=[on|off] to enable/disable asserts on demand

### PR DESCRIPTION
**Pull request by mathias-lang-sociomantic**
FYI @MartinNowak @mihails-strasuns-sociomantic

### CODE REVIEW

**Comment by Geod24** [comment]
Hum, silly me - That can't work because of bootstrapping. @MartinNowak : How are you compiling the releases ? Could we do this in 2 steps, by first introducing this switch, and enabling it later on ?

File: src/mars.d [hunk 1]
```diff
@@ -358,20 +359,24 @@ private int tryMain(size_t argc, const(char)** argv)
     getenv_setargv(readFromEnv(&environment, "DFLAGS"), &arguments);
     updateRealEnvironment(&environment);
     environment.reset(1); // don't need environment cache any more
-    version (none)
```
Comment (WalterBright) [inline src/mars.d]: This is for debugging purposes. Please leave it in.

**Comment by WalterBright** [comment]
I don't think this adds much of any value to `-release`. Note that `-release` has the following effects:
```
if (global.params.release)
{
    global.params.useInvariants = false;
    global.params.useIn = false;
    global.params.useOut = false;
    global.params.useAssert = false;
    global.params.useSwitchError = false;
}
if (global.params.useUnitTests)
    global.params.useAssert = true;
```

**Comment by WalterBright** [comment]
It's also becoming clear that turning asserts on and off with the command line is a bit blunt, especially when projects tend to be all files on one command line. Some asserts are more equal than others. Individual asserts can be left on with:
```
if (!condition) assert(0);
```
but writing code that way is awkward and just doesn't look good.

I propose instead to do something analogous to the accepted `pragma(inline, true/false)`, such as:
```
pragma(assert); // default
pragma(assert, true); // turn asserts on
pragma(assert, false); // turn asserts off
```
Can you write a DIP for it?

**Comment by mathias-lang-sociomantic** [comment]
> I don't think this adds much of any value to `-release`.

Yeah, we could allow more fine-grained control over what is enabled / disabled, but in any case, it won't solve the problem it's trying to solve, since we need to compile with the oldest C++ DMD.
Instead of changing D behavior, I'm considering just replacing all `assert` calls with `throw new IceError()`... Would that work, or do you have other suggestions ?

**Comment by WalterBright** [comment]
> Would that work

I don't see how it would be any different.

`-release` enables several checks, as I listed above. Probably a decent step is to try turning them on/off individually, and see which one contributes most to the 8% slowdown.

**Comment by mathias-lang-sociomantic** [comment]
> I don't see how it would be any different.

Well the problem it would introduce is that we won't be able to use this new feature directly for a while.
And as we keep compatibility with the last C++ DMD, it only makes matter worse.
Using an `Error` type would solve that issue.

Regarding `pragma(assert)`, I'm not sure I understand your position completely.
It would allow one function (/ module / aggregate) to force `assert` to stay, right ? The point of this switch is to unconditionally do that for all module, rather than allowing fine-grained control (which is already possible by other means, such as the `assert(0)` trick you described, or a manual `throw Error`).

**Comment by mathias-lang-sociomantic** [comment]
The issue itself (15962) needs another fix.
The functionality (more control over `assert` than what `debug` / `release` provide) will be covered by https://github.com/dlang/DIPs/pull/54

**Pull request closed without merging**